src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,197 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+#if DEBUG
+using System.Diagnostics;
+#endif
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    internal partial class TemporaryStorageServiceFactory
+    {
+        private sealed class ReferenceCountedDisposable<T> : IDisposable
```

Comment:
> If we are implementing a ref-counting scheme that depends on Dispose being called

It does not depend on `Dispose` being called. It's more efficient that way, but it will not fail to release resources if used sub-optimally. (This is the same set of conditions as the previous implementation.)

In addition, we know that some callers will fail to call `Dispose` in the current implementation. (Also the same as before.)

> It seems very easy for someone to make one of these types, potentially forget to dispose it, and then leak a refcount, leading to never letting go of the underlying file.

It will not leak. (the following explanation will be added in a comment)

The underlying object is _eligible_ for collection when all references to the same object reach _either_ of the following states:

1. The reference is eligible for collection (no path to root)
1. The reference is disposed

The underlying object will be _deterministically_ disposed when all references to the same object are explicitly disposed.